### PR TITLE
chore: migrate trial details [DET-4021]

### PR DESCRIPTION
**Pull request by hamidzr**
## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: proto/src/determined/trial/v1/trial.proto [hunk 1]
```diff
@@ -34,7 +36,15 @@ message CheckpointWorkload {
 // MetricsWorkload is a workload generating metrics.
 message MetricsWorkload {
   option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
-    json_schema: { required: [ "start_time", "state" ] }
+    json_schema: {
+      required: [
+        "startTime",
```
Comment (hamidzr) [inline proto/src/determined/trial/v1/trial.proto:41]: https://determinedai.atlassian.net/browse/DET-4648

**Comment by hkang1** [comment]
**chore:** the table row entries need a `key`
Screen Shot 2020-12-10 at 12 48 39 PM

File: webui/react/src/utils/types.ts [hunk 2]
```diff
@@ -156,7 +166,7 @@ export const oneOfProperties = <T>(obj: any, props: string[]): T => {
 };
 
 // size in bytes
-export const checkpointSize = (checkpoint: Checkpoint): number => {
+export const checkpointSize = (checkpoint: Checkpoint | CheckpointWorkload): number => {
```
Comment (hkang1) [inline webui/react/src/utils/types.ts:169]: **suggestion:** what do you think about moving trial specific type utilities into `utils/trial`? `utils/types` is starting to get a bit large

File: webui/react/src/utils/trial.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ export const extractMetricValue = (step: Step, metricName: MetricName): number |
   return undefined;
 };
 
-export const extractMetricNames = (steps: Step[] = []): MetricName[] => {
+export const extractMetricNamesStep = (steps: Step[] = []): MetricName[] => {
```
Comment (hkang1) [inline webui/react/src/utils/trial.ts:17]: **nit:** `extractMetricNamesFromSteps`

File: webui/react/src/utils/trial.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
-import { MetricName, MetricType, Step } from 'types';
+import { MetricName, MetricType, Step, WorkloadWrapper } from 'types';
 import { isNumber, metricNameSorter } from 'utils/data';
 
 import handleError, { DaError, ErrorLevel, ErrorType } from '../ErrorHandler';
 
-export const extractMetricValue = (step: Step, metricName: MetricName): number | undefined => {
+export const extractMetricValueStep = (step: Step, metricName: MetricName): number | undefined => {
```
Comment (hkang1) [inline webui/react/src/utils/trial.ts:6]: **nit:** `extractMetricValuesFromSteps`

File: webui/react/src/utils/types.ts (hunk 2 above)
Comment (hkang1) [reply webui/react/src/utils/types.ts:169]: we can leave that work to be a separate ticket. Basic idea of breaking up `utils/types` into the respective type it is dealing with.

File: webui/react/src/utils/step.ts [hunk 5]
```diff
@@ -1,5 +1,41 @@
...
+: workload is MetricsWorkload => {
+  return 'metrics' in workload
+  && 'numInputs' in workload
+  && !('uuid' in workload)
+  && !('resources' in workload);
+};
+
+export const workloadsToSteps = (workloads: WorkloadWrapper[]): Step2[] => {
+  const stepsDict: Record<number, Partial<Step2>> = {};
+  workloads.forEach(wlWrapper => {
```
Comment (hkang1) [inline webui/react/src/utils/step.ts:26]: **nit:** `workload` instead of `wlWrapper`?

File: webui/react/src/types.ts [hunk 6]
```diff
@@ -237,22 +238,49 @@ export interface MetricName {
...
+}
+
 // Validation sub step.
 export interface Validation extends StartEndTimes {
   id: number;
   state: RunState;
   metrics?: ValidationMetrics;
 }
 
+export interface Step2 extends StartEndTimes {
```
Comment (hkang1) [inline webui/react/src/types.ts:277]: **question:** not sure if this is any better:
```
export interface Step2 extends WorkloadWrapper, StartEndTimes {
```

**Review by hkang1 — COMMENTED** [review]
This looks really good! Just the small table `key` issue. Other than that it should be good to go!

File: proto/src/determined/trial/v1/trial.proto (hunk 1 above)
Comment (hamidzr) [reply proto/src/determined/trial/v1/trial.proto:41]: that has now landed. I'll take this out.

File: webui/react/src/types.ts (hunk 6 above)
Comment (hamidzr) [reply webui/react/src/types.ts:277]: aah I thought TS wouldn't be happy with overwriting the `training` attribute but it didn't complain. this is better

File: webui/react/src/utils/step.ts (hunk 5 above)
Comment (hamidzr) [reply webui/react/src/utils/step.ts:26]: :thinking:  maybe I should have used wrapper in the original function parameter too. I add wrapper to be more explicit about this not being a `Workload` type

File: webui/react/src/utils/types.ts (hunk 2 above)
Comment (hamidzr) [reply webui/react/src/utils/types.ts:169]: in general that sounds good. this specific one is also used in the checkpoint modal module and isn't specific to trials.

File: webui/react/src/utils/step.ts (hunk 5 above)
Comment (hamidzr) [reply webui/react/src/utils/step.ts:26]: how about renaming workloads to wlWrappers or wrappers? or workloadWrappers

File: webui/react/src/utils/types.ts (hunk 2 above)
Comment (hamidzr) [reply webui/react/src/utils/types.ts:168]: https://github.com/determined-ai/determined/pull/1674/commits/f34d5ce2bf2dd94c95872adba1ab03919cdc284a

**Review by hkang1 — APPROVED** [review]
Nice the key error is gone! 🚢 🎈

**Pull request merged**